Fix template error handling; add logging support

portante opened the pull request:
The exceptions were being caught in the wrong order such that `ConnectionError` exceptions, which are a sub-class of `TransportError`, were not being considered.

We also add support for only updating a template when the template's mapping version is different.

We add a `logger` argument for `streaming_bulk()` so that we have a bit more insight into what is happening while processing actions.

Further, we fix the failing unit tests, add Travis CI integration, and ensure the code passes flake8.

maulikjs commented:
Hey Folks,
Do we have an update on this PR?

portante commented:
Just waiting for a code-review ...

dbutenhof commented on pyesbulk/__init__.py:
```diff
@@ -68,23 +101,58 @@ def put_template(es, name, body):
+        body_ver = int(body["mappings"][mapping_name]["_meta"]["version"])
+    except KeyError as e:
+        raise Exception(
+            f"Bad template, {name}, given mapping name misssing from template,"
+            f" {mapping_name}: {e!r}"
```
Well, OK; but potentially misleading since it could be any of the four key values.

portante replied:
Fixed.

dbutenhof commented on pyesbulk/__init__.py:
```diff
@@ -68,23 +101,58 @@ def put_template(es, name, body):
+        except es_excs.TransportError as exc:
+            # Only retry on certain 500 errors
+            if exc.status_code != 404:
+                if exc.status_code not in [500, 503, 504]:
```
Does Elasticsearch document certain 500 errors as retry-able? Normally, 500 errors are a sort of boggy nowhere since the server is telling you it doesn't know what happened.

Alternatively, since this is a GET operation, which had better be idempotent, there seems little harm in retrying an unknown error at least once.

dbutenhof commented on pyesbulk/__init__.py:
```diff
@@ -68,23 +101,58 @@ def put_template(es, name, body):
+                    tmpl[name]["mappings"][mapping_name]["_meta"]["version"]
+                )
+            except KeyError:
+                pass
```
Really? This indicates something's corrupted in the document we got back from Elasticsearch, which I think would demand at least a log entry.

dbutenhof commented on pyesbulk/__init__.py:
```diff
@@ -68,23 +101,58 @@ def put_template(es, name, body):
+            _sleep_w_backoff(backoff)
             backoff += 1
             retry_count += 1
         else:
```
And we don't worry about the "re-try-able" 500 errors here at all? Granted, probably rare, but since we're putting a known document to replace a known name, seems a retry would still be idempotent?

portante replied:
With the op-type of "create" this is okay, because we ensure the caller is providing the document IDs, which means we'll get a duplicate back on re-try if we received a 500 error, but somehow the document was also indexed.

For the "update" op-type, this is also okay, because the document will just get updated again, but the expected error if the document ID does not already exist.

For the "delete" op-type, this is also okay, because if it was deleted, we'll just get a 404 back.

For the "index" op-type, if the ID already exists, it will be updated again, and if it doesn't already exist it will be created.

Seems like this is a good comment to add...

portante replied:
> And we don't worry about the "re-try-able" 500 errors here at all?

Can you explain more what you mean?  This code will retry all 500, 503, and 504 errors.  Any other 5xx error re-raise the `TransportError` exception.

What do we have to worry about when retrying PUT operations on a template?

dbutenhof replied:
Ha ha; somehow I completely missed that the check was repeated here. Weird.

Although maybe it'd be better to capture the non-retryable list in a CONSTANT so it's immediately clear we're using the same set in both places.

portante replied:
Fixed.

dbutenhof commented on pyesbulk/__init__.py:
```diff
@@ -183,25 +281,74 @@ def actions_tracking_closure(cl_actions):
+                try:
+                    exc_payload = resp["exception"]
+                except KeyError:
+                    pass
```
Should we get some meaningful exc_payload here instead of just ignoring the KeyError? Otherwise we can't distinguish resp["exception"] == None from lack of resp["exception"] ... which I can imagine being a useful distinction in weird cases.

portante replied:
I'll add the following comment which may help:

```
                    # We have an exception object in the response object
                    # which is not always JSON serializable, so we use
                    # `repr` to turn that exception into a serializable
                    # string while maintaining as much information about
                    # the exception as possible.
```

dbutenhof replied:
My concern was whether the value of exc_payload in the `else:` below would give us a useful clue as to whether `resp["exception"]` == None, or was missing, because the `pass` seems to silently obscure that distinction.

portante replied:
So here are the three cases we are trying to handle:

 1. the key `exception` is not in the `resp` dictionary -- do nothing, as we have no object to ensure it is JSON serializable
 2. the key `exception` is in the `resp` dictionary -- emit a `repr` of whatever value is contained under that dictionaries key
 3. the key `exception` is in the `resp` dictionary but is `None` -- using `repr` should emit the string `'None'`

Does that make it clear regarding `pass` behavior?  Or am I still missing what you are trying to communicate?

dbutenhof commented on pyesbulk/__init__.py:
```diff
@@ -183,25 +281,74 @@ def actions_tracking_closure(cl_actions):
+                    },
+                    indent=4,
+                    sort_keys=True,
+                )
```
Random JSON dumps in the log file are hard to interpret; it'd be nice to have a bit of context, too.

portante replied:
Actually, these are not really random JSON dumps.  They contain the response object, and the original "action" dictionary that was sent.  The intent is to have this captured in a file, not sent in some log message, so that the caller can decide what to do with actions that were indexed.

Most of the time, the response object has enough context, combined with the original action payload, so that one can do something about it.

dbutenhof replied:
My point was that you're writing a JSON dump into a file without any situation/location context. Someone familiar with the code may recognize where that would have been written and why, but seeing just the logged output would give anyone else no clue at all what it means.

ndokos commented on tests/put_template_test.py:
```diff
@@ -14,11 +14,13 @@ def __init__(self, mock):
 
-class MockPutTemplate(object):
-    def __init__(self, behavior=None):
+class MockTemplateApis(object):
```
Leaving out `object` is the new normal, right?

ndokos commented on pyesbulk/__init__.py:
```diff
@@ -122,30 +191,37 @@ def streaming_bulk(es, actions, errorsfp):
-            # if after yielding an action some actions appear on the retry deque
-            # start yielding those actions until we drain the retry queue.
+            # If after yielding an action some actions appear on the retry
+            # deque start yielding those actions until we drain the retry
```
"deque" --> "deque,"

portante replied:
Fixed.

ndokos commented on pyesbulk/__init__.py:
```diff
@@ -157,21 +233,43 @@ def actions_tracking_closure(cl_actions):
+            status = resp["status"]
+        except KeyError as e:
+            assert not ok, f"ok = {ok!r}, e = {e!r}"
+            # Limit the length of the error message.
```
Is this a FIXME comment? I don't see how the length of the message is limited.

portante replied:
Fixed.

ndokos commented on tests/put_template_test.py:
```diff
@@ -14,11 +14,13 @@ def __init__(self, mock):
 
-class MockPutTemplate(object):
-    def __init__(self, behavior=None):
+class MockTemplateApis(object):
```
Not sure what happpened to my earlier comment, so here goes again: Leaving out `object` is the new normal, right?

ndokos replied:
Well, I see it above, but not while in review...

portante replied:
Yes, in Python3 you are no longer required to explicitly sub-class `object`.

npalaska commented on pyesbulk/__init__.py:
```diff
@@ -43,65 +68,115 @@ def _calc_backoff_sleep(backoff):
+            # We retry all connection errors
+            _sleep_w_backoff(backoff)
+            backoff += 1
+            retry_count += 1
```
Does it make sense to add a condition on the retry count as well along with a boolean retry?

portante replied:
Can you explain more of what you mean?  What behavior would you be changing / fixing by doing that?

portante replied:
@npalaska?

dbutenhof commented:
Hey, Pete; while you're touching __init__.py here, consider adding the __VERSION__ constant that's required by your setup.py!

Otherwise I might push a PR that does it, but I'm not sure whether you intend this to be "1.0" or something less grand.

With that change (and a pipenv shell for the setup builder dependencies, because I can't seem to make it happy with the global configuration) I can build a wheel that actually installs. (Whee.)

portante commented:
> Hey, Pete; while you're touching `__init__.py` here, consider adding the `__VERSION__` constant that's required by your `setup.py`!

Fixed!

> With that change (and a `pipenv` shell for the setup builder dependencies, because I can't seem to make it happy with the global configuration) I can build a wheel that actually installs. (Whee.)

Also added `python3 setup.py sdist bdist_wheel` as a TravisCI step as well.

dbutenhof reviewed (approved):
Relatively minor stuff that can be cleaned up later.

portante commented on pyesbulk/__init__.py:
```diff
@@ -157,21 +256,43 @@ def actions_tracking_closure(cl_actions):
+        except KeyError as e:
+            assert not ok, f"ok = {ok!r}, e = {e!r}"
+            # Limit the length of the error message.
+            logger.error("{!r}"[:_MAX_ERRMSG_LENGTH], e)
```
You are right, this message truncation will not work this way at all!  (grumbles, mutters, ...)

portante commented on pyesbulk/__init__.py:
```diff
@@ -43,65 +72,128 @@ def _calc_backoff_sleep(backoff):
+            if exc.status_code == 404:
+                # We expected a "not found", we'll PUT below.
+                pass
+            elif exc.status_code < 500:
```
Good point.

portante commented on .travis.yml:
```diff
@@ -0,0 +1,13 @@
+---
+language: python
+install:
+  - pip install flake8 wheel
```
Probably can't hurt to lock in at 3.8.3 for consistency.

portante commented on pyesbulk/__init__.py:
```diff
@@ -183,25 +304,79 @@ def actions_tracking_closure(cl_actions):
+                    pass
+                else:
+                    resp["exception"] = repr(exc_payload)
+                try:
+                    error = resp["error"]
+                except KeyError:
+                    error = ""
```
Here I think the subtle difference is that `error` is used in the `if` condition, where the `resp['exception']` value is just a payload fixup.

So we care that `error` is a valid string so that `error.startswith()` functions properly, but we don't care if `resp['exception']` exists or not, it is just dumped into a JSON string properly with the fixup we applied if it does exist.

dbutenhof commented on pyesbulk/__init__.py:
```diff
@@ -285,8 +288,7 @@ def actions_tracking_closure(cl_actions):
             assert not ok, f"ok = {ok!r}, e = {e!r}"
-            # Limit the length of the error message.
-            logger.error("{!r}"[:_MAX_ERRMSG_LENGTH], e)
+            logger.error("{!r}", e)
```
So back to no truncation at all? How big can a KeyError payload get, though?

portante replied:
I am not sure why I added it in the first place.  It can't get that big.
